modules/ballerina-http/src/main/java/org/ballerinalang/net/http/HttpUtil.java

```diff
@@ -448,12 +454,40 @@ public static void startPendingHttpConnectors() throws BallerinaConnectorExcepti
         }
     }
 
-    public static void handleResponse(HTTPCarbonMessage requestMsg, HTTPCarbonMessage responseMsg) {
+    public static BValue[] prepareResponseAndSend(Context context, AbstractNativeFunction abstractNativeFunction
+            , HTTPCarbonMessage requestMessage, HTTPCarbonMessage responseMessage) {
+        addHTTPSessionAndCorsHeaders(requestMessage, responseMessage);
+        HttpResponseStatusFuture statusFuture = handleResponse(requestMessage, responseMessage);
         try {
-            requestMsg.respond(responseMsg);
+            statusFuture = statusFuture.sync();
+        } catch (InterruptedException e) {
+            throw new BallerinaException("interrupted sync: " + e.getMessage());
+        }
+        if (statusFuture.getStatus() != null) {
```

Comment:
Shall we change the API to always return state which can contain both success and failure state.

Reply:
ack